proof/refine/ARM/Reply_R.thy

```diff
@@ -95,4 +109,462 @@ lemma replyRemoveTCB_st_tcb_at'_sym_ref:
+  unfolding updateReply_def
+  apply wpsimp
+  apply (erule rsubst2[where P=P])
+   apply (clarsimp simp: ext get_refs_def opt_map_def list_refs_of_reply'_def obj_at'_def
+                         projectKO_eq
+                  split: option.split)+
+  done
+
+lemma updateReply_replyPrev_Nothing[wp]:
+  "\<lbrace>\<lambda>s. P (replyNexts_of s) ((replyPrevs_of s)(rptr := Nothing))\<rbrace>
+   updateReply rptr (replyPrev_update (\<lambda>_. Nothing))
+   \<lbrace>\<lambda>_ s. P (replyNexts_of s) (replyPrevs_of s)\<rbrace>"
+  unfolding updateReply_def
+  apply wpsimp
+  apply (erule rsubst2[where P=P])
+   apply (clarsimp simp: ext get_refs_def opt_map_def list_refs_of_reply'_def obj_at'_def
+                         projectKO_eq
+                  split: option.split)+
+  done
+
+lemma cleanReply_list_refs_of_replies':
+  "\<lbrace>\<lambda>s. P ((list_refs_of_replies' s)(rptr := {}))\<rbrace>
+   cleanReply rptr
+   \<lbrace>\<lambda>_ s. P (list_refs_of_replies' s)\<rbrace>"
+  unfolding cleanReply_def decompose_list_refs_of_replies'
+  apply wpsimp
+  apply (erule rsubst[where P=P])
+  apply (rule ext)
+  apply (clarsimp simp: list_refs_of_reply'_def fun_upd_def)
+  done
+
+lemma isHead_to_head:
+  "isHead x = (\<exists>scPtr. x = Some (Head scPtr))"
+  "(\<forall>scPtr. y \<noteq> Head scPtr) = (\<exists>rPtr. y = Next rPtr)"
+  apply (clarsimp simp: isHead_def split: option.split reply_next.split)
+  apply (case_tac y; clarsimp)
+  done
+
+(* FIXME RT: remove. proved by MikiT *)
+lemma sym_refs_replyNext_replyPrev_proj:
+  "sym_refs (list_refs_of_replies' s') \<Longrightarrow>
+    replyNexts_of s' rp = Some rp' \<longleftrightarrow> replyPrevs_of s' rp' = Some rp"
+  sorry
+
+(* FIXME RT: remove. proved by MikiT *)
+lemma sym_refs_replyNext_None:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s');
+    replyNexts_of s' rp = None \<rbrakk> \<Longrightarrow> \<forall>rp'. replyPrevs_of s' rp' \<noteq> Some rp"
+  sorry
+
+(* FIXME RT: remove. proved by MikiT *)
+lemma sym_refs_replyPrev_None:
+  "\<lbrakk>sym_refs (list_refs_of_replies' s');
+    replyPrevs_of s' rp = None \<rbrakk> \<Longrightarrow> \<forall>rp'. replyNexts_of s' rp' \<noteq> Some rp"
+  sorry
+
+lemma ks_reply_at'_repliesD:
+  "\<lbrakk>replies_of' s rptr = Some reply; sym_refs (list_refs_of_replies' s)\<rbrakk>
+   \<Longrightarrow> replyNexts_of s rptr = replyNext_of reply
+       \<and> (case replyNext_of reply of
+            Some next \<Rightarrow> replyPrevs_of s next = Some rptr
+          | None \<Rightarrow> \<forall>rptr'. replyPrevs_of s rptr' \<noteq> Some rptr)
+       \<and> replyPrevs_of s rptr = replyPrev reply
+       \<and> (case replyPrev reply of
+            Some prev \<Rightarrow> replyNexts_of s prev = Some rptr
+          | None \<Rightarrow> \<forall>rptr'. replyNexts_of s rptr' \<noteq> Some rptr)"
+  \<comment>\<open> opt_mapE will sometimes destroy the @{term "(|>)"} inside @{term replyNexts_of}
```

Comment:
I don’t quite understand this comment... are you sure this `supply` line is doing anything?

Reply:
Just reiterating our conversation on mattermost: `opt_mapE` is declared `[elim!]` somewhere, and that declaration messes up the proof state here. Turns out we remove it from `[elim!]` on RISCV64; [VER-1367](https://sel4.atlassian.net/browse/VER-1367) asks if we should just remove it from `[elim!]` everywhere.

Separately, you mentioned you got a copy of this proof to go through without the `supply` line - I'll see if that works for me too.